src/tss2-fapi/ifapi_ima_eventlog.c

```diff
@@ -0,0 +1,853 @@
+/* SPDX-License-Identifier: BSD-2 */
+/*******************************************************************************
+ * Copyright 2017-2018, Fraunhofer SIT sponsored by Infineon Technologies AG
+ * All rights reserved.
+ ******************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <openssl/sha.h>
+#include <ctype.h>
+#include <stdbool.h>
+#include <openssl/evp.h>
+#include <json-c/json.h>
+#include <json-c/json_util.h>
+#include "tss2_common.h"
+#include "tss2_tpm2_types.h"
+#include "fapi_int.h"
+#include "ifapi_json_deserialize.h"
+#include "tpm_json_deserialize.h"
+#include "ifapi_ima_eventlog.h"
+#include "ifapi_helpers.h"
+
+#define LOGMODULE fapijson
+#include "util/log.h"
+#include "util/aux_util.h"
+
+/* Defines from kernel ima.h */
+#define IMA_TEMPLATE_FIELD_ID_MAX_LEN 16
+#define IMA_TEMPLATE_NUM_FIELDS_MAX 15
+/* Define from kernel crypt.h */
+#define CRYPTO_MAX_ALG_NAME 128
+
+static TSS2_RC
+get_json_content(json_object *jso, json_object **jso_sub) {
+    if (!ifapi_get_sub_object(jso, CONTENT, jso_sub)) {
+        *jso_sub = json_object_new_object();
+        return_if_null(*jso_sub, "Out of memory.", TSS2_FAPI_RC_MEMORY);
+        json_object_object_add(jso, CONTENT, *jso_sub);
+    }
+    return TSS2_RC_SUCCESS;
+}
+
+static TSS2_RC
+add_uint8_ary_to_json(UINT8 *buffer, UINT32 size, json_object *jso, const char *jso_tag)
+{
+    TSS2_RC r;
+    json_object *jso_byte_string = NULL;
+
+    return_if_null(buffer, "Bad reference.", TSS2_FAPI_RC_BAD_VALUE);
+    return_if_null(jso, "Bad reference.", TSS2_FAPI_RC_BAD_VALUE);
+
+    char *hex_string = malloc((size) * 2 + 1);
+    return_if_null(hex_string, "Out of memory.", TSS2_FAPI_RC_MEMORY);
+
+    if (size > 0) {
+        for (size_t i = 0, off = 0; i < size; i++, off += 2)
+            sprintf(&hex_string[off], "%02x", buffer[i]);
+    }
+    hex_string[(size) * 2] = '\0';
+    jso_byte_string = json_object_new_string(hex_string);
```

Comment:
insert `SAFE_FREE(hex_string);` here and you can avoid the goto and just return if error.